simplified-reader-bookmarks/src/main/java/org/nypl/simplified/books/reader/bookmarks/internal/RBService.kt

```diff
@@ -0,0 +1,320 @@
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarkEvent
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarkHTTPCallsType
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarkServiceType
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarkSyncEnableResult
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarkSyncEnableResult.SYNC_DISABLED
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarkSyncEnableResult.SYNC_ENABLED
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarkSyncEnableResult.SYNC_ENABLE_NOT_SUPPORTED
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarkSyncEnableStatus
+import org.nypl.simplified.reader.bookmarks.api.ReaderBookmarks
+import org.slf4j.LoggerFactory
+import java.util.Collections
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+class RBService(
+  private val threads: (Runnable) -> Thread,
+  private val httpCalls: ReaderBookmarkHTTPCallsType,
+  private val bookmarkEventsOut: Subject<ReaderBookmarkEvent>,
+  private val profilesController: ProfilesControllerType
+) : ReaderBookmarkServiceType {
+
+  private val disposables =
+    CompositeDisposable()
+
+  private val executor: ListeningScheduledExecutorService =
+    MoreExecutors.listeningDecorator(
+      Executors.newScheduledThreadPool(1) { runnable ->
+        RBServiceThread(this.threads.invoke(runnable))
+      }
+    )
+
+  private val logger =
+    LoggerFactory.getLogger(RBService::class.java)
+  private val objectMapper =
+    ObjectMapper()
+  private val accountsSyncChanging =
+    Collections.synchronizedSet(hashSetOf<AccountID>())
+
+  init {
+    this.disposables.add(
+      this.profilesController.profileEvents()
+        .subscribe(this::onProfileEvent)
+    )
+    this.disposables.add(
+      this.profilesController.accountEvents()
+        .subscribe(this::onAccountEvent)
+    )
+
+    /*
+     * Sync bookmarks hourly.
+     */
+
+    this.executor.scheduleAtFixedRate({ this.onRegularSyncTimeElapsed() }, 0L, 1L, TimeUnit.HOURS)
+  }
+
+  private fun onProfileEvent(event: ProfileEvent) {
+    if (event is ProfileSelection.ProfileSelectionInProgress) {
+      try {
+        val currentProfile = this.profilesController.profileCurrent()
+        this.logger.debug("[{}]: a new profile was selected", currentProfile.id.uuid)
+        this.onProfileWasSelected()
+      } catch (e: ProfileNoneCurrentException) {
+        this.logger.error("onProfileEvent: no profile is current")
+      }
+    }
+  }
+
+  private fun onAccountEvent(event: AccountEvent) {
+    if (event is AccountEventLoginStateChanged) {
+      when (event.state) {
+        is AccountLoggedIn ->
+          this.onAccountLoggedIn()
+
+        is AccountLoggingIn,
+        is AccountLoggingInWaitingForExternalAuthentication,
+        is AccountLoggingOut,
+        is AccountLoginFailed,
+        is AccountLogoutFailed,
+        AccountNotLoggedIn -> {
```

Comment:
Isn't a `is` missing here?

Reply:
`AccountNotLoggedIn` is actually an `object`, so you can't refer to it in an `is` expression. Kotlin oddity.

Reply:
Got it! I thought it was a state too